presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java

```diff
@@ -99,15 +105,84 @@
     private final Metadata metadata;
     private final DeterminismEvaluator determinismEvaluator;
 
+    private final LoadingCache<RowExpression, Supplier<PageProjection>> projectionCache;
+    private final LoadingCache<RowExpression, Supplier<PageFilter>> filterCache;
+
+    private final CacheStatsMBean projectionCacheStats;
+    private final CacheStatsMBean filterCacheStats;
+
     @Inject
-    public PageFunctionCompiler(Metadata metadata)
+    public PageFunctionCompiler(Metadata metadata, CompilerConfig config)
+    {
+        this(metadata, config.isExpressionCacheEnabled());
+    }
+
+    public PageFunctionCompiler(Metadata metadata, boolean expressionCacheEnabled)
     {
         this.metadata = requireNonNull(metadata, "metadata is null");
         this.determinismEvaluator = new DeterminismEvaluator(metadata.getFunctionRegistry());
+
+        if (expressionCacheEnabled) {
+            projectionCache = CacheBuilder.newBuilder().recordStats().maximumSize(10_000).build(
+                    new CacheLoader<RowExpression, Supplier<PageProjection>>()
+                    {
+                        @Override
+                        public Supplier<PageProjection> load(RowExpression projection)
+                                throws Exception
+                        {
+                            return compileProjectionInternal(projection, Optional.empty());
+                        }
+                    });
+            projectionCacheStats = new CacheStatsMBean(projectionCache);
+        }
+        else {
+            projectionCache = null;
+            projectionCacheStats = null;
+        }
+
+        if (expressionCacheEnabled) {
+            filterCache = CacheBuilder.newBuilder().recordStats().maximumSize(10_000).build(
+                    new CacheLoader<RowExpression, Supplier<PageFilter>>()
+                    {
+                        @Override
+                        public Supplier<PageFilter> load(RowExpression projection)
+                                throws Exception
+                        {
+                            return compileFilterInternal(projection, Optional.empty());
+                        }
+                    });
+            filterCacheStats = new CacheStatsMBean(filterCache);
+        }
+        else {
+            filterCache = null;
+            filterCacheStats = null;
+        }
+    }
+
+    @Managed
+    @Nested
+    public CacheStatsMBean getProjectionCache()
+    {
+        return projectionCacheStats;
+    }
+
+    @Managed
+    @Nested
+    public CacheStatsMBean getFilterCache()
+    {
+        return filterCacheStats;
     }
 
     public Supplier<PageProjection> compileProjection(RowExpression projection, Optional<String> classNameSuffix)
     {
+        if (projectionCache == null) {
+            return compileProjectionInternal(projection, classNameSuffix);
+        }
+        return projectionCache.getUnchecked(projection);
```

Comment:
The suffix is a hint to the compiler.  We recently added this so when the cache is disabled you get the queryId and plan node in the class name, and thus, the stack traces.

Also we are not sure the cache is even needed.  This came up from a very small benchmark sf100 on a 100 node cluster, but reasonable sized workloads don't need a cache.  The plan is to add this and then test in some real world workloads.